### DISCUSSION TITLE
작성자가 본인인지 체크하는 로직의 중복을 줄이기 위해 어떻게 할까요?

### DISCUSSION THREAD

**Discussion by mobzzzzz** [one-day-one-meal/one-hour-one-meal#47] [Ideas]
우선 본격적인 CRUD 작성을 시작하고 예외처리까지 꼼꼼히 작성해주셔서 감사합니다!

기왕 하는거 코드의 중복을 줄이는 방법도 고민해보고 있는데 의견 여쭙습니다

```
// User Entity
fun hasPermission(requestUserId: Long, roles: String): Boolean {
    return this.id == requestUserId || roles == "ROLE_ADMIN"
}

// User Service
check(
    !it.hasPermission(principal.id, principal.authorities.firstOrNull().toString())
) { "No permission to access" }
```
            
            
            
TMI: check() 는 Kotlin이 제공해주는 Preconditions로 value가 false면 IllegalStateException("$message")를 발생시킵니다.

우선 **Entity는 고결해야한다** 라는 키워드에 걸맞게 Entity 에서 UserPrincipal 을 Param에 사용하면 클린 아키텍처적으로 좋지 않다고 해서 이런 모양새가 나옵니다.

그리고 인가에 대한 체크는 사실 Entity가 할법한 일이 아닌 느낌이라 Entity에서 분리하면 결국 서비스에서 해결한다면 다음과 같이 될 것 같습니다

```
// 어딘가의 함수
fun checkPermission(entityOwnerId: Long, principal: Principal) {
      val hasPermission = (entityOwnerId == principal.id || principal.authorities.firstOrNull().toString() == "ROLE_ADMIN")
      check(hasPermission) { "User does not have permissions" }
}

// 어딘가의 서비스
findByIdOrNull()
?.also { PermissionChecker.checkPermission(it.user.id, principal) } // Id 체크
?.let { ... }
```

이걸 Component로 만들어서 주입받는 것도 애매하고 그렇다고 object로 만들어서 뺄까 싶자니 좀 단순해서 고민이 되네요.

서비스 로직에서의 체크 로직에 대한 여러분의 의견이 궁금합니다.

**Comment by devitssu** [comment]
ROLE은 `Method Security`를 사용하면 간단히 해결 할 수 있을거 같은데 작성자 체크까지 하고싶다면 별도로 작성하긴 해야겠네요. `CustomPermissionEvaluator`를 만드는 방법이 있는것 같습니다.
https://www.baeldung.com/spring-security-create-new-custom-security-expression

**Comment by mobzzzzz** [comment]
#48 에서 제가 생각하는 로직 본인 확인 공통 처리부분 구현해봤습니다 나중에 튜터님께 배워올게요

**Comment by mobzzzzz** [comment]
튜터님 피드백 맞춰서 Entity 에서 Permission check 에 대한 책임을 가져가게 바꾸겠습니다.
